src/Test.elm

```diff
@@ -99,7 +100,21 @@ describe untrimmedDesc tests =
                 , reason = Test.Expectation.Invalid Test.Expectation.EmptyList
                 }
         else
-            Internal.Labeled desc (Internal.Batch tests)
+            case Internal.duplicatedName tests of
+                Err duped ->
+                    Internal.failNow
+                        { description = "The tests '" ++ desc ++ "' contains multiple tests named '" ++ duped ++ "'. Do some renaming so that tests have unique names."
+                        , reason = Test.Expectation.Invalid Test.Expectation.DuplicatedName
+                        }
+
+                Ok childrenNames ->
+                    if Set.member desc childrenNames then
+                        Internal.failNow
+                            { description = "The test '" ++ desc ++ "' contains a child test of the same name. Do some renaming so that tests have distinct names."
+                            , reason = Test.Expectation.Invalid Test.Expectation.DuplicatedName
+                            }
+                    else
+                        Internal.Labeled desc (Internal.Batch tests)
```

Comment:
I think `Test.concat` needs to run `duplicatedName` too.

https://github.com/elm-community/elm-test/pull/133/files#r103121485 would fail if it's only on `describe`!